File: Journey/Level_0_fundamentals/vector.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int main() {
    int n;
    cin >> n;

    vector<int> v1, v2;
    v1.reserve(n);
    v2.reserve(n);

    for (int i = 0; i < n; i++) {
        int x;
        cin >> x;
        v1.push_back(x);
    }

    for (int i = 0; i < n; i++) {
        int y;
        cin >> y;
        v2.push_back(y);
    }

    vector<int> sum(n), diff(n);
    int dp = 0;

    for (int i = 0; i < n; i++) {
        sum[i] = v1[i] + v2[i];
        diff[i] = v1[i] - v2[i];
        dp += v1[i] * v2[i];
    }


    float l2_norm_v1 = 0;

    for(int i = 0; i < v1.size(); i++){
        l2_norm_v1 = l2_norm_v1 + v1[i] * v1[i];
    } 
   
    l2_norm_v1 = sqrt(l2_norm_v1);

    float l2_norm_v2 = 0;

    for(int i = 0; i < v2.size(); i++){
        l2_norm_v2 = l2_norm_v2 + v2[i] * v2[i];
    } 
     l2_norm_v2 = sqrt(l2_norm_v2);

    float cosine_similarity_v1_v2 = 0;

    if(l2_norm_v1 == 0 || l2_norm_v2 == 0){
        cosine_similarity_v1_v2 = 0;
    }else{
        cosine_similarity_v1_v2 = (dp)/(l2_norm_v1 * l2_norm_v2);
    }


    cout << "Sum: ";
    for (int x : sum) cout << x << " ";
    cout << "\n";

    cout << "Difference: ";
    for (int x : diff) cout << x << " ";
    cout << "\n";

    cout << "Dot Product: " << dp << "\n";
    cout << "\n";

    cout<<"L2_Norm of V1 is : " <<l2_norm_v1 <<"\n";
    cout<<"\n";
 
    cout<<"L2_Norm of V2 is : " <<l2_norm_v2 <<"\n";
    cout<<"\n";

    cout<<"cosine similarity between v1 and v2 is : " <<cosine_similarity_v1_v2<<"\n";
    cout<<"\n";


    return 0;
}

File: Journey/Level_0_fundamentals/vector.py
'''

Task:  Write a program that takes two vectors of equal length

Computes: Element-wise sum, Element wise difference, Dot product, L2 Norm


The L2 norm, or Euclidean norm, measures a vector's length (magnitude) as the straight-line distance from the origin, calculated by squaring each component, summing them up, and taking the square root.


'''

import numpy as np 

import math

v1 = np.array(list(map(float, input("Enter vector 1: ").split())))

v2 = np.array(list(map(float, input("Enter vector 2: ").split())))

if len(v1)!= len(v2):
    print("length of two vectors should be same, Please check ")
    exit()


summ = v1 + v2
diff = v1 - v2
dp = np.dot(v1, v2)

l2_norm_v1 = 0
for ele in v1:
    l2_norm_v1 = l2_norm_v1 + ele * ele
l2_norm_v1 = math.sqrt(l2_norm_v1)

l2_norm_v2 = 0
for ele in v2:
    l2_norm_v2 = l2_norm_v2 + ele * ele
l2_norm_v2 = math.sqrt(l2_norm_v2)


cosine_similarity_v1_v2 = 0

if l2_norm_v2 == 0 or l2_norm_v1 == 0 :
    cosine_similarity_v1_v2 = 0
else:  
    cosine_similarity_v1_v2 = dp/(l2_norm_v1 * l2_norm_v2)

print("Element-wise Sum:", summ)
print("Element-wise Difference:", diff)
print("Dot Product:", dp)
print("L2 norm of vector 1 is : ", l2_norm_v1)
print("L2 norm of vector 2 is : ", l2_norm_v2)
print("Cosine Similarity between vectors V1 and V2 is : ", cosine_similarity_v1_v2)

